data-prepper-plugins/mutate-event-processors/src/main/java/org/opensearch/dataprepper/plugins/processor/mutateevent/CopyValueProcessor.java

```diff
@@ -13,41 +13,84 @@
-                }
+            try {
+                final Event recordEvent = record.getData();
+                if (config.getFromList() != null || config.getToList() != null) {
+                    // Copying entries between lists
+                    if (recordEvent.containsKey(config.getToList()) && !config.getOverwriteIfToListExists()) {
+                        continue;
+                    }
 
-                if (entry.getFromKey().equals(entry.getToKey()) || !recordEvent.containsKey(entry.getFromKey())) {
-                    continue;
-                }
+                    final List<Map<String, Object>> sourceList = recordEvent.get(config.getFromList(), List.class);
+                    final List<Map<String, Object>> targetList = new ArrayList<>();
+
+                    final Map<CopyValueProcessorConfig.Entry, Boolean> whenConditions = new HashMap<>();
+                    for (final CopyValueProcessorConfig.Entry entry : entries) {
+                        if (Objects.nonNull(entry.getCopyWhen()) && !expressionEvaluator.evaluateConditional(entry.getCopyWhen(), recordEvent)) {
+                            whenConditions.put(entry, Boolean.FALSE);
+                        } else {
+                            whenConditions.put(entry, Boolean.TRUE);
+                        }
+                    }
+                    for (final Map<String, Object> sourceField : sourceList) {
+                        final Map<String, Object> targetItem = new HashMap<>();
+                        for (final CopyValueProcessorConfig.Entry entry : entries) {
+                            if (!whenConditions.get(entry) || !sourceField.containsKey(entry.getFromKey())) {
+                                continue;
+                            }
+                            targetItem.put(entry.getToKey(), sourceField.get(entry.getFromKey()));
+                        }
+                        targetList.add(targetItem);
+                    }
+                    recordEvent.put(config.getToList(), targetList);
+                } else {
+                    // Copying individual entries
+                    for (final CopyValueProcessorConfig.Entry entry : entries) {
+                        if (Objects.nonNull(entry.getCopyWhen()) && !expressionEvaluator.evaluateConditional(entry.getCopyWhen(), recordEvent)) {
+                            continue;
+                        }
+
+                        if (entry.getFromKey().equals(entry.getToKey()) || !recordEvent.containsKey(entry.getFromKey())) {
+                            continue;
+                        }
 
-                if (!recordEvent.containsKey(entry.getToKey()) || entry.getOverwriteIfToKeyExists()) {
-                    final Object source = recordEvent.get(entry.getFromKey(), Object.class);
-                    recordEvent.put(entry.getToKey(), source);
+                        if (!recordEvent.containsKey(entry.getToKey()) || entry.getOverwriteIfToKeyExists()) {
```

Comment:
Suggest refactor L77-85 into a private method `shouldCopyEntry` or `skipCopyEntry`

Reply:
Good one. Made the change.

Reply:
Also opened an issue for adding metrics: https://github.com/opensearch-project/data-prepper/issues/4088